### PR TITLE
Trois maréchaux

### PR DESCRIPTION
**Pull request by eggrobin**
avec mes sabots.

### CODE REVIEW

File: ksp_plugin_adapter/utf8_marshaler.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    return instance_;
+  }
+
+  public override void CleanUpNativeData(IntPtr native_data) {
+    Marshal.FreeHGlobal(native_data);
+#if MARSHAL_DEBUG
+    // Useful for debugging things from a console application to see whether
+    // allocating and freeing match.  We cannot use Log.Info here, since that
+    // would recursively call the marshaler and end in overflowing hilarity.
+    Console.WriteLine("freeh " + Convert.ToString(native_data.ToInt64(), 16));
```
Comment (pleroy) [inline ksp_plugin_adapter/utf8_marshaler.cs]: Director of the FBI?

File: ksp_plugin_adapter/utf8_marshaler.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  public override IntPtr MarshalManagedToNative(object managed_object) {
+    var value = managed_object as String;
+    if (value == null) {
+      throw Log.Fatal(String.Format(CultureInfo.InvariantCulture,
+                                    "|{0}| must be used on a |{1}|.",
+                                    GetType().Name,
+                                    typeof(String).Name));
+    }
+    int size = utf8_.GetByteCount(value);
+    IntPtr ptr = Marshal.AllocHGlobal(size + 1);
```
Comment (pleroy) [inline ksp_plugin_adapter/utf8_marshaler.cs]: ptr->buffer?

File: ksp_plugin_adapter/utf8_marshaler.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  public override void CleanUpNativeData(IntPtr native_data) {}
+  public override IntPtr MarshalManagedToNative(object managed_object) {
+    throw Log.Fatal("use |InUTF8Marshaler| for in parameters");
+  }
+
+  public override unsafe object MarshalNativeToManaged(IntPtr native_data) {
+    sbyte* begin = (sbyte*)native_data;
+    sbyte* end;
```
Comment (pleroy) [inline ksp_plugin_adapter/utf8_marshaler.cs]: Declare end in the loop.

File: ksp_plugin_adapter/utf8_marshaler.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    for (end = begin; *end != 0; ++end) {}
+    return new String(begin, 0, (int)(end - begin), utf8_);
+  }
+
+  private readonly static OutUTF8Marshaler instance_ = new OutUTF8Marshaler();
+}
+
+// A marshaler for out parameter or return value UTF-8 strings whose ownership
+// is taken by the caller.
+internal class UTF8FactoryMarshaler : OutUTF8Marshaler {
```
Comment (pleroy) [inline ksp_plugin_adapter/utf8_marshaler.cs]: The class name is obscure.  OutOwnedUTF8Marshaler?

File: ksp_plugin_adapter/utf8_marshaler.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  public static new ICustomMarshaler GetInstance(String s) {
+    return instance_;
+  }
+
+  public override void CleanUpNativeData(IntPtr native_data) {
+    // NOTE(egg): this should get renamed.
```
Comment (pleroy) [inline ksp_plugin_adapter/utf8_marshaler.cs]: You don't say.

File: ksp_plugin_adapter/utf8_marshaler.cs (hunk 4 above)
Comment (eggrobin) [inline ksp_plugin_adapter/utf8_marshaler.cs]: Done.

File: ksp_plugin_adapter/utf8_marshaler.cs (hunk 2 above)
Comment (eggrobin) [inline ksp_plugin_adapter/utf8_marshaler.cs]: Done.

**Pull request merged**